it/google-cloud-platform/src/main/java/org/apache/beam/it/gcp/dataflow/AbstractPipelineLauncher.java

```diff
@@ -79,7 +86,7 @@ public Job getJob(String project, String region, String jobId) throws IOExceptio
 
   @Override
   public Job getJob(String project, String region, String jobId, String view) {
-    return Failsafe.with(clientRetryPolicy())
+    return Failsafe.with(clientRetryPolicy(), dataflowRetryPolicy())
```

Comment:
429 is what already happens, retry will help deflake, but on the other hand may cause the quota more stressed.

I assume the most likely call returning 429 is get job here. We periodically fetch the job. Is it possible not retry here and when HTTP Error 429 happens, just assemble and return an empty job, with unknown status. Later this will return UNKNOWN status:

https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/100a410066ca02da31918fcff54724f5dff1f8bc/it/google-cloud-platform/src/main/java/org/apache/beam/it/gcp/dataflow/AbstractPipelineLauncher.java#L248

Then the launcher will retry fetch job in the next 15 seconds.

Reply:
Typically it is at the end when updating the job, i.e. when cancelling or draining

Reply:
Example:
```
2024-06-25T18:44:04.0391954Z PUT https://dataflow.googleapis.com/v1b3/projects/cloud-teleport-testing/locations/us-central1/jobs/2024-06-25_11_36_34-1035710696993183123
2024-06-25T18:44:04.0393387Z {
2024-06-25T18:44:04.0393754Z   "code": 429,
2024-06-25T18:44:04.0394152Z   "details": [
2024-06-25T18:44:04.0394606Z     {
2024-06-25T18:44:04.0395264Z       "@type": "type.googleapis.com/google.rpc.ErrorInfo",
2024-06-25T18:44:04.0396038Z       "reason": "RATE_LIMIT_EXCEEDED"
2024-06-25T18:44:04.0396560Z     },
2024-06-25T18:44:04.0396880Z     {
2024-06-25T18:44:04.0397456Z       "@type": "type.googleapis.com/google.rpc.Help"
2024-06-25T18:44:04.0398045Z     }
2024-06-25T18:44:04.0398366Z   ],
2024-06-25T18:44:04.0398714Z   "errors": [
2024-06-25T18:44:04.0399072Z     {
2024-06-25T18:44:04.0399475Z       "domain": "global",
2024-06-25T18:44:04.0401708Z       "message": "Quota exceeded for quota metric 'Job update requests' and limit 'Job update requests per minute per user' of service 'dataflow.googleapis.com' for consumer 'project_number:269744978479'.",
2024-06-25T18:44:04.0403494Z       "reason": "rateLimitExceeded"
2024-06-25T18:44:04.0404002Z     }
2024-06-25T18:44:04.0404322Z   ],
2024-06-25T18:44:04.0406406Z   "message": "Quota exceeded for quota metric 'Job update requests' and limit 'Job update requests per minute per user' of service 'dataflow.googleapis.com' for consumer 'project_number:269744978479'.",
2024-06-25T18:44:04.0408142Z   "status": "RESOURCE_EXHAUSTED"
2024-06-25T18:44:04.0408605Z }
```

Reply:
The method you linked just checks the status, it does not tap into the 'Job update requests' metric unless I'm mistaken. I just added the retry there in case it does hit another 429 issue, but I could just remove that if there is worry.